src/Language/Printer.php

```diff
@@ -70,256 +80,330 @@ public static function doPrint($ast)
-                        ? $selectionSet
-                        : $this->join([$op, $this->join([$name, $varDefs]), $directives, $selectionSet], ' ');
-                },
-                NodeKind::VARIABLE_DEFINITION => function(VariableDefinitionNode $node) {
-                    return $node->variable . ': ' . $node->type . $this->wrap(' = ', $node->defaultValue);
-                },
-                NodeKind::SELECTION_SET => function(SelectionSetNode $node) {
-                    return $this->block($node->selections);
-                },
-                NodeKind::FIELD => function(FieldNode $node) {
-                    return $this->join([
-                        $this->wrap('', $node->alias, ': ') . $node->name . $this->wrap('(', $this->join($node->arguments, ', '), ')'),
-                        $this->join($node->directives, ' '),
-                        $node->selectionSet
-                    ], ' ');
-                },
-                NodeKind::ARGUMENT => function(ArgumentNode $node) {
-                    return $node->name . ': ' . $node->value;
-                },
+        return Visitor::visit(
+            $ast,
+            [
+                'leave' => [
+                    NodeKind::NAME                      => function (Node $node) {
+                        return '' . $node->value;
+                    },
+                    NodeKind::VARIABLE                  => function ($node) {
+                        return '$' . $node->name;
+                    },
+                    NodeKind::DOCUMENT                  => function (DocumentNode $node) {
+                        return $this->join($node->definitions, "\n\n") . "\n";
+                    },
+                    NodeKind::OPERATION_DEFINITION      => function (OperationDefinitionNode $node) {
+                        $op           = $node->operation;
+                        $name         = $node->name;
+                        $varDefs      = $this->wrap('(', $this->join($node->variableDefinitions, ', '), ')');
+                        $directives   = $this->join($node->directives, ' ');
+                        $selectionSet = $node->selectionSet;
+                        // Anonymous queries with no directives or variable definitions can use
+                        // the query short form.
+                        return ! $name && ! $directives && ! $varDefs && $op === 'query'
+                            ? $selectionSet
+                            : $this->join([$op, $this->join([$name, $varDefs]), $directives, $selectionSet], ' ');
+                    },
+                    NodeKind::VARIABLE_DEFINITION       => function (VariableDefinitionNode $node) {
+                        return $node->variable . ': ' . $node->type . $this->wrap(' = ', $node->defaultValue);
+                    },
+                    NodeKind::SELECTION_SET             => function (SelectionSetNode $node) {
```

Comment:
Usually there's a way how to format the code better rather than ignoring the CS rule :) Check the last commit, I tried to separate the blocks. To me, they make better sense like this.

Reply:
The new version is OK for me. Thanks!

P.S. I used to write the code with paddings like in this code style, but it was such a pain when you renamed something, that I abandoned it. Suprised that Doctrine uses it, but I guess I can live with it with a help of IDE and auto fix.

Reply:
PS: same for me. But it was the time when it was hard to hide whitespace in diffs (no such feature in github) etc. 

But it seems to make code blocks to look well-aranged to me so I adopted it again :) Tools make it easy to use.